libconsensus/pbft/PBFTEngine.cpp

```diff
@@ -919,96 +921,102 @@ void PBFTEngine::checkAndCommit()
+                                 << LOG_KV("reqIdx", m_reqCache->prepareCache()->idx)
+                                 << LOG_KV(
+                                        "hash", m_reqCache->prepareCache()->block_hash.abridged())
+                                 << LOG_KV("nodeIdx", nodeIdx())
+                                 << LOG_KV("myNode", m_keyPair.pub().abridged())
+                                 << LOG_KV("genSigTimeCost", genSig_time_cost)
+                                 << LOG_KV("commitBlockTimeCost", commitBlock_time_cost)
+                                 << LOG_KV("dropTxsTimeCost", dropTxs_time_cost)
+                                 << LOG_KV("noteSealingTimeCost", noteSealing_time_cost)
+                                 << LOG_KV("totalTimeCost", utcTime() - start_commit_time);
+            m_reqCache->delCache(m_reqCache->prepareCache()->block_hash);
+            m_reqCache->removeInvalidFutureCache(m_highestBlock);
         }
         else
         {
             PBFTENGINE_LOG(WARNING)
-                << LOG_DESC("checkAndSave: Consensus Failed, Block already exists")
-                << LOG_KV("reqNum", m_reqCache->prepareCache()->height)
+                << LOG_DESC("CommitBlock Failed")
+                << LOG_KV("reqNum", p_block->blockHeader().number())
                 << LOG_KV("curNum", m_highestBlock.number())
-                << LOG_KV("blkHash", m_reqCache->prepareCache()->block_hash.abridged())
-                << LOG_KV("highHash", m_highestBlock.hash().abridged())
+                << LOG_KV("reqIdx", m_reqCache->prepareCache()->idx)
+                << LOG_KV("hash", p_block->blockHeader().hash().abridged())
                 << LOG_KV("nodeIdx", nodeIdx()) << LOG_KV("myNode", m_keyPair.pub().abridged());
+            /// note blocksync to sync
+            m_blockSync->noteSealingBlockNumber(m_blockChain->number());
+            m_txPool->handleBadBlock(*p_block);
         }
     }
+    else
+    {
+        PBFTENGINE_LOG(WARNING)
+            << LOG_DESC("checkAndCommitBlock: Consensus Failed, Block already exists")
+            << LOG_KV("reqNum", m_reqCache->prepareCache()->height)
+            << LOG_KV("curNum", m_highestBlock.number())
+            << LOG_KV("blkHash", m_reqCache->prepareCache()->block_hash.abridged())
+            << LOG_KV("highHash", m_highestBlock.hash().abridged()) << LOG_KV("nodeIdx", nodeIdx())
+            << LOG_KV("myNode", m_keyPair.pub().abridged());
+    }
+}
+
+/// if collect >= 2/3 SignReq and CommitReq, then callback this function to commit block
+/// check whether view and height is valid, if valid, then commit the block and clear the context
+void PBFTEngine::checkAndSave()
+{
+    size_t sign_size = m_reqCache->getSigCacheSize(m_reqCache->prepareCache()->block_hash);
+    size_t commit_size = m_reqCache->getCommitCacheSize(m_reqCache->prepareCache()->block_hash);
+    if (sign_size >= minValidNodes() && commit_size >= minValidNodes())
```

Comment:
do not call minValidNdoes() twice~
```suggestion
    auto min_nodes = minValidNodes()
    if (sign_size >= min_nodes && commit_size >= min_nodes)
```

Reply:
done, pls refer to: https://github.com/FISCO-BCOS/FISCO-BCOS/pull/827/files#diff-b7e4695daa9e7f6fd6da2fc48711cd6cR1017